### PR TITLE
Proper signal handling

### PR DESCRIPTION
**Pull request by s0undt3ch**
REVIEW CAREFULLY!!!!

### CODE REVIEW

**Comment by s0undt3ch** [comment]
Whoops... some fixes required.

**Comment by cachedout** [comment]
Yup, looks like we've still got some issues here. :]

**Comment by s0undt3ch** [comment]
Yeah, the code changes regarding the master signals handling was not _compatible_ with making the master's `start()` method run on a separate process(as opposed to instantiating and running the master on a separate process). With the latest changes, it should actually run the tests now.

**Comment by s0undt3ch** [comment]
The lint failures are unrelated to my changes...

**Comment by s0undt3ch** [comment]
Err. Tests are still not running on Jenkins....

**Comment by jtand** [comment]
Go Go Jenkins!

**Comment by s0undt3ch** [comment]
Go Go Jenkins!

**Comment by s0undt3ch** [comment]
Go Go Jenkins!

**Comment by s0undt3ch** [comment]
Go Go Jenkins!

**Comment by s0undt3ch** [comment]
Go Go Jenkins!

**Comment by s0undt3ch** [comment]
Go Go Jenkins!

**Comment by s0undt3ch** [comment]
Go Go Jenkins!

**Comment by s0undt3ch** [comment]
Go Go Jenkins!

**Comment by s0undt3ch** [comment]
Go Go Jenkins!

**Comment by s0undt3ch** [comment]
/cc @jacksontj can you please check the commits where I implemented the deep copy dunder and see if they don't make any sense or are bluntly wrong?

Whenever we try to deep copy an object which has an ioloop as an attribute it will fail because the ioloop contains a thread lock which is not deep copiable...

File: salt/crypt.py [hunk 1]
```diff
@@ -377,6 +378,19 @@ def __singleton_init__(self, opts, io_loop=None):
         else:
             self.authenticate()
 
+    def __deepcopy__(self, memo):
+        cls = self.__class__
+        result = cls.__new__(cls, copy.deepcopy(self.opts, memo), io_loop=None)
+        memo[id(self)] = result
+        for key, value in self.__dict__.items():
```
Comment (jacksontj) [inline salt/crypt.py]: If you are going to do deepcopy of the value, I'm not certain you need .items() (as that will make a shallow copy as well).

**Comment by jacksontj** [comment]
The code looks alright, although I'm curious under what circumstances do we deepcopy all of these channels? That seems like something we shouldn't be doing ;)

**Comment by s0undt3ch** [comment]
https://jenkins.saltstack.com/job/salt-linode-cent5/2169/console the reason why those 2 commits were added...

On further investigation and extra debugging it all went down to the threading Lock that io loops have which can't be deep copied. So, either we don't store references to the io_loop's and just get the current one(if that does not break current functionality) or, we implement `__deepcopy__`...

**Comment by s0undt3ch** [comment]
@jacksontj: To make it clear, you're stack context changes is what's causing all that copying.... :]

**Comment by s0undt3ch** [comment]
*your

**Comment by s0undt3ch** [comment]
To add more to the clarification, the offending instances were found in `__context__`.... should we sanitize the context in case we can't deep copy it instead?

**Comment by s0undt3ch** [comment]
Go Go Jenkins!

**Comment by s0undt3ch** [comment]
The tests should still pass, I added some commits which ended in https://github.com/saltstack/salt/pull/28697 and did a hard-reset to this branch'es PR leaving it on the last know commit(which was passing).

**Comment by s0undt3ch** [comment]
Go Go Jenkins!

**Pull request merged**